[Python3] Use correct Python version for LLDB-related tests

tbkka opened the pull request:
LLDB builds a Python module in the form of a C shared library.  This is used in some Swift tests.

Since Python module format changed between Python 2 and Python 3, these tests must be run with the same Python version that was used when LLDB built it.

Of course, in a system with multiple Python versions installed, this may or may not be the same version used by the rest of the Swift build.

This adds some logic to lit.cfg to figure out the version of Python used by LLDB and expose that as `%{lldb-python}` for use in tests that need to access the LLDB module.

tbkka committed:
Use the correct Python executable to match LLDB

The linux-fatal-backtrace script needs to run with an LLDB
module loaded into Python.  This in turn requires that the
test be run with the exact same Python version as was used
by LLDB (not just the same Python module directory).  This
can get confused on systems with multiple versions of Python
installed.

This replaces `lldb-python-path` (the Python module directory path)
with `lldb-python` (which is the correct Python version run with
the LLDB path).  This should ensure that this test is always
run with the same Python version and module that LLDB used.

tbkka committed:
Use consistent braces for the lldb-python substitution

tbkka commented:
@swift-ci Please test Linux

tbkka commented:
Since the only affected test runs only on Linux, let's fully test there first and make sure everything works before running a smoke test on macOS.

compnerd commented on test/lit.cfg:
```diff
@@ -104,6 +104,12 @@ def get_lldb_python_path(lldb_build_root):
     return subprocess.check_output([lldb_path, "-P"]).rstrip().decode('utf-8')
 
+def get_lldb_python_from_path(lldb_path):
+    for c in lldb_path.split('/'):
```
Could you use `os.sep` instead of `/` please?

tbkka replied:
Even better, I used `os.path.split` to separate the path into components (which required rewriting this to a recursive structure instead of iterative).  That should work correctly on all platforms, even those where there are multiple valid path separators.

compnerd reviewed (approved):
Thanks, this definitely will allow the tests to pass.  However, we should also simultaneously look at migrating the lldb build to python3.  Either way, we should do this to make the system less fragile.

tbkka committed:
Use os.path utilities to dissect paths

tbkka commented:
@swift-ci Please test Linux

tbkka committed:
Allow `python3` as a path identifier

Previous code required a decimal point before it would
recognize a path component as a Python version identifier,
so it would accept `python2.7` but not `python3`.

tbkka commented:
@swift-ci Please smoke test macOS

rintaro commented:
@swift-ci Please smoke test Linux

JDevlieghere commented:
While this is fine as a ad-hoc solution, really what we want here is to figure out the Python LLDB was build against from the LLDB build because it's the one-and-only source of truth. I can envision many scenarios where this won't work, the simplest one where a user passes a -DPYTHON_HOME CMake flag to LLDB to a Python that's not in path.

tbkka commented:
If the `_lldb` module is intended to be used by user Python scripts, then there needs to be a mechanism outside of the build itself for doing this.  Currently, the `lldb` executable has the `-P` option to provide the path to the built module.  There should be another option to provide a path to a suitable python executable that can run that module.  (I suppose LLDB itself could serve as that executable, since it has the appropriate Python version embedded in it. 🤔)

JDevlieghere commented:
> If the `_lldb` module is intended to be used by user Python scripts, then there needs to be a mechanism outside of the build itself for doing this.

I disagree, I don't think that's our responsibility. Nothing prevents you from running Python from inside the embedded script interpreter, maybe we should modify the test to do that instead? FWIW this constraint is not limited to LLDB but applies to anyone that builds a Python module that itself links against Python. 

Also, the potential problem I described above is inherent to the unfortunate `swift` <- `lldb` <- `swift test` dependency chain. If this were an LLDB test this wouldn't be an issue because the build would be aware of which Python was used. This is not really different from making sure LLVM and Swift are using the same Python interpreter. But because of the dependency we can't really tell you. I guess another solution would be for us to say copy the used Python interpreter into a well known location (say next to the lldb binary) and have the Swift test use that.

JDevlieghere commented:
My preferred solution would be to move to @compnerd  unified build!

JDevlieghere commented:
Implemented this in #33235
